IET-based parameters and alternative argument "engine"

mlange05 opened the pull request:
#### Reviewer note:
 Apologies for the somewhat large PR, but unfortunately certain assumptions in the current `parameters` implementation entail that both, parameters and arguments, have to be changed at the same time. There are also still a few minor "hacks" and shortcuts that we might want to iron out during or after thorough review, so I'm submitting this now.

### Description:
This PR has two primary objectives: 
* Change the derivation of required operator/function inputs (`.parameters`) to be based on the actual IR/IET of the function. This entails we now only require `parameters` that are actually used by the function/operator, instead of the conservative over-estimate that we previously used.
* Replace the current `ArgumentEngine` with an (objectively)  much cleaner symbol-based mechanism for deriving argument values. The new mechanism uses two stages: default value derivation (`argument_defaults`) and runtime value overrides based on user input (`argument_values`), both of which are defined by the symbol classes themselves. This makes the definition of our "derivation rules" for arguments more accessible and allows things like aliasing to be implemented much more easily. 
_Sidenote: The new mechanism encapsulates the same shallow tree/forest-like traversal pattern as the previous "engine" but in a more natural way, as the intrinsic relationships (eg. `Composite->Function->Dimension->Scalar`) are already defined by each symbol (eg. a `Function` knows it's `Dimension`s (`indices`) and each `Dimension` knows its `start_symbolic` or its `parent`)._

### TLDR; Detailed feature/change list
**Part 1: IET-based parameter derivation**
* Use `Scalar` types for iteration bound variables and add them to the `free-symbols` finder utility.
* Add a new symbol finder mode `defines` that collects symbols locally defined symbols by an IET node.
* Utility to derive `parameters` as the set of `functions/symbolics` + `free-symbols` - `defines`.
* Add explicit `ArrayCast` and `PointerCast` nodes to the tree to pick up the correct `dim_size` variables required by casts.
* Use new `derive_parameters` utility to derive argument/parameter list for elemental functions.
_Note: This part is probably still a bit wonky and can be simplified further, either as part of the review discussion or in a follow-up PR._

**Part 2: Symbol-based argument derivation**
* No more standalone `ArgumentEngine`
* `parameters` are defined as a list of symbolic objects and arguments are defined as a `name->value` map. There are no more dedicated `xxxArgument` types. 
* Each `Function/Dimension` type now provides `argument_default` and `argument_values(**kwargs)` to derive default values and user-defined overrides.
* The defaulting mechanism utilises a modified `MultiDict` to collect candidate values from multiple sources (eg. multiple `Function` objects may provide values for `Dimension` sizes). This `ArgumentDict` then performs a candidate reduction (`unique()`) during retrieval to ensure the value is uniquely defined and throws an error if competing values are present.
_Note: This could be extended with additional reduction rules if we need such "tie breaker" mechanisms, but the current code base does not actually require it._
* `SteppingDimension`: Derived dimension variables do not show up in `parameters` anymore, as they are now defined locally. This means each `DerivedDimension` object effectively adds their candidate values to the `.parent` during default and value derivation, a.k.a "derived dims alias their parents".
* For value overrides, `op(dim=<value>, ...)` now aliases to `op(dim_e=<value>, ...)` and contributes candidate values to the respective map entry.
* User-defined values for `CompositeFunction` or `Function` types now can easily use `new.default_values(...)` to re-derive default values like dimension sizes during the override stage.

#### Final notes:
* Argument derivation should probably also include a third stage `argument_validation` that checks the default + override values against actual data shapes to pick up inconsistencies. Since the PR is already huge and I'm kind of out of time, I've left this for a follow-up PR (not necessarily by me).
* ~Travis is still failing for one last set of tests all related to `example_gradient.py`. As the regular gradient test is performing fine, I'm assuming this is only a minor inconsistency somewhere in the test and would like to go ahead with the review for now, as it is quite comprehensive.~ Fixed now.

fixes #465 #416 #387

FabioLuporini commented on devito/types.py:
```diff
@@ -165,6 +165,22 @@ def function(self):
     def indexify(self):
         return self
 
+    def argument_defaults(self):
```
why not in ``Basic``, or in a Mixin?

mlange05 replied:
Sure, why not. :+1:

FabioLuporini commented on devito/types.py:
```diff
@@ -345,6 +361,22 @@ def indexify(self, indices=None):
         indices = [a.subs(subs) for a in self.args]
         return Indexed(self.indexed, *indices)
 
+    def argument_defaults(self):
```
if in ``Basic``, as I wrote above, we can drop some redundant code

mlange05 replied:
:+1:

FabioLuporini commented on tests/test_autotuner.py:
```diff
@@ -101,7 +101,7 @@ def test_timesteps_per_at_run():
                      outfield.indexify() + infield.indexify()*3.0)
         op = Operator(stencil, dle=('blocking', {'blockalways': True}))
-        op(infield=infield, outfield=outfield, autotune=True)
+        op(infield=infield, outfield=outfield, t=2, autotune=True)
```
?

mlange05 replied:
Ahhh, this is a subtle one. In the previous scheme, a stepping dimension would provide its buffer size as the default runtime loop bound. Since this makes no sense from a modellers perspective, this rule got dropped, meaning every time-dependente operator now needs to either use `save=some_value` or define the number of timesteps on invocation. This test does not (because it doesn't care), but this now gets picked up as a missing value for `time`, hence we need to provide one.

FabioLuporini replied:
OK

FabioLuporini replied:
> Ahhh, this is a subtle one. [...] hence we need to provide one.

I had to apply the same fix to another couple of tests

mlange05 commented on devito/ir/iet/nodes.py:
```diff
@@ -156,6 +157,24 @@ def __init__(self, name, params=None):
         return "Call::\n\t%s(...)" % self.name
 
+    @property
+    def functions(self):
```
:+1:

mlange05 commented on devito/ir/iet/nodes.py:
```diff
@@ -156,6 +157,24 @@ def __init__(self, name, params=None):
+    def free_symbols(self):
+        """Return all :class:`Symbol` objects used by this :class:`Call`."""
+        free = tuple(set(flatten(p.free_symbols for p in self.params)))
+        # HACK: Filter dimensions to avoid them on popping onto outer parameters
```
Ok, so the example arises when elementals encapsulate the `x` and `y` loops. In this case, `x` and `y` still appear in `functions` due to being part of the symbol `u(t, x, y)`, but the `Iteration` that `defines` them is in an elemental. Now, they are undefined to the outer `Operator` and thus appear on the argument list, unless explicitly taken out. As stated, it's a bit of a hack...

Also :+1: on the `cached_property`.

FabioLuporini replied:
and I guess you tried fixing this issue within `_create_elemental_functions`, without success, right?

anyway, I read your description 10 times and I'm still unsure what's actually going on. Is this the situation:

```
void elemental (...) {
  ...
}

void kernel (... int x, int y ...) {
   ...
   elemental ( ... )
}
```
? or what?

mlange05 replied:
Yes, exactly. Nothing in the main `Operator` actually defines the symbols (which makes sense), but the presence of the `Function` symbol brings `x`/`y` into `arguments.

mlange05 commented on devito/ir/iet/nodes.py:
```diff
@@ -297,6 +328,14 @@ def __repr__(self):
+        """
+        Return any symbols an :class:`Iteration` may define.
+        """
+        dim = self.dim.parent if self.dim.is_Stepping else self.dim
```
:+1:

mlange05 commented on devito/ir/iet/nodes.py:
```diff
@@ -419,6 +458,23 @@ def children(self):
 
+    @property
+    def functions(self):
+        """Return all :class:`Function` objects used by this :class:`Iteration`"""
```
:+1:

mlange05 commented on devito/ir/iet/utils.py:
```diff
@@ -119,3 +119,21 @@ def compose_nodes(nodes, retrieve=False):
         return body
+
+
+def derive_parameters(nodes):
```
I believe it already does, no?

FabioLuporini replied:
right, I guess I was looking at something else

mlange05 commented on devito/dle/backends/basic.py:
```diff
@@ -68,43 +64,41 @@ def _create_elemental_functions(self, nodes, state):
+                    args.append((defined_args[i.name], i))
                 elif i.is_Array:
-                    args.append(("(%s*)%s" % (c.dtype_to_ctype(i.dtype), i.name), i))
+                    args.append((i, i))
```
:+1:

mlange05 commented on devito/ir/iet/visitors.py:
```diff
@@ -132,34 +131,59 @@ def _args_decl(self, args):
-            if i.is_ScalarArgument:
+            if i.is_Object:
+                ret.append(c.Value('void', '*_%s' % i.name))
+            elif i.is_Scalar:
```
:+1

mlange05 replied:
Actually no, doesn't work. Turns out objects that are `is_Scalar` do not have the `.is_Lowered` attribute, as they are not `Dimension` types. This then causes errors, so I'll have to leave this one as it is.

mlange05 commented on devito/operator.py:
```diff
@@ -85,7 +86,8 @@ def __init__(self, expressions, **kwargs):
 
         # Parameters of the Operator (Dimensions necessary for data casts)
-        parameters = self.input + self.dimensions
+        parameter_dims = [d for d in self.dimensions if not d.is_Stepping]
```
This code is actually dead now, as the actual arguments are overwritten later. I'll remove the dead code.

FabioLuporini replied:
excellent

FabioLuporini commented on devito/ir/iet/nodes.py:
```diff
@@ -156,6 +157,24 @@ def __init__(self, name, params=None):
+    @property
+    def functions(self):
+        """Return all :class:`Symbol` objects used by this :class:`Call`."""
+        return tuple(p for p in self.params if isinstance(p, types.AbstractFunction))
```
``... if p.is_AbstractFunction`` ?

mlange05 replied:
Nope, the DLE insert actual expressions into the call signature, `self.params` is not guaranteed to be internal symbols only.

FabioLuporini replied:
OK

FabioLuporini commented on devito/ir/iet/nodes.py:
```diff
@@ -156,6 +157,24 @@ def __init__(self, name, params=None):
+
+    @property
+    def free_symbols(self):
+        """Return all :class:`Symbol` objects used by this :class:`Call`."""
```
same docstring as ``def functions`` ?

mlange05 replied:
:+1:

mlange05 commented on devito/operator.py:
```diff
@@ -117,24 +119,66 @@ def __init__(self, expressions, **kwargs):
+        # Derive parameters as symbols not defined in the kernel itself
+        parameters = derive_parameters(nodes)
+
+        # Filter all internally-allocated temporary `Array` types
```
Well, not really as all the DLE transformations have to happen for the IET to be in its final shape and the whole point of this is to use the actual final IET and not lug around assumptions from a previous IR state.

mlange05 replied:
But parameter derivation is about more than just function collection - it requires knowledge of who defines (all types of) variables. This means, all things that adjust variable definitions, like elementals or `insert_C_decls` have to have happened before we can safely derive the final `parameter`s. 

What this "hack" is really doing is effectively "predicting" what the `Allocator` will do, because `Allocator` only inserts raw CGen format for local temporary allocations, which cannot be read from the IET state. In general, I don't like this approach of "predicting what another part of the code will do", since this is exactly what caused the inaccurate over-estimation of parameters in the first place (and thus got us into this mess). So instead, I think "the right thing to do" here is to adjust the `Allocator` to insert local array allocations into the IET, allowing us to actually use a fixed state for derivation instead of a "previous state + assumptions".

mlange05 commented on devito/operator.py:
```diff
@@ -117,24 +119,66 @@ def __init__(self, expressions, **kwargs):
-        self.argument_engine = ArgumentEngine(clusters.ispace, parameters,
-                                              self.dle_arguments)
+        # Insert data and pointer casts for array parameters and profiling structs
+        functions = FindSymbols('symbolics').visit(nodes)
```
Yes, it would seem so. I also note that `self.input` does not seem to be use anywhere, but should probably be let for another day.

mlange05 commented on devito/operator.py:
```diff
@@ -117,24 +119,66 @@ def __init__(self, expressions, **kwargs):
+        casts = [ArrayCast(f) for f in functions if f.is_Tensor and f._mem_external]
+        profiler = Object(self.profiler.name, self.profiler.dtype, self.profiler.new)
+        casts.append(PointerCast(profiler))
+        nodes = (List(body=casts), nodes)
```
Hmm, I actually tried generalising this and moving it to a utility function, because the same thing needs to be done for newly created elementals. However, this is not as easy as it looks, due to the funky way the profiler `Object`s are handled. Since elementals don't need profiling, I think it would be great to simply pick the `Object` off the `TimedList` IET nodes, but that soon became messy. I'll give this another go, but maybe in a separate PR?

mlange05 commented on devito/operator.py:
```diff
@@ -117,24 +119,66 @@ def __init__(self, expressions, **kwargs):
+
+        # Derive additional values for DLE arguments
+        # TODO: This is not pretty, but it works for now. Ideally, the
+        # DLE arguments would be massaged into the IET so as to comply
```
See my comment about IET state and `Allocator` above. I've tried this, but it turned out to be riddled with subtleties.

FabioLuporini replied:
OK

mlange05 commented on devito/operator.py:
```diff
@@ -117,24 +119,66 @@ def __init__(self, expressions, **kwargs):
+        # First, derive all default values from parameters and reduce
+        # TODO: Should be self.parameters, but that converts to
+        # TensorArgument type, etc...
+        # Also, this part can probably be a cached_property
```
Looks like this "just works". Nice. :smile:

mlange05 commented on devito/operator.py:
```diff
@@ -171,9 +215,11 @@ def cfunction(self):
             argtypes = []
             for i in self.parameters:
-                if i.is_ScalarArgument:
+                if i.is_Object:
```
Since `Object`s don't adhere to the internal hierarchy they don't have `is_Scalar` or `is_Tensor`, so we need to catch them first to avoid errors.

FabioLuporini replied:
really? how is that? Doesn't ``Object`` inherit from ``Basic`` ?

mlange05 commented:
Ok, I believe I have addressed all/most of @FabioLuporini s comments.

I've also fixed the `gradient_example`-related test failures. Turns out this "second implementation" of our base test uses a lot of weird symbol-name aliasing (eg. `op(v=TimeFunction(name='u', ...), ...)` or `op(rec=SparseFunction(name='rec_g', ...), ...)`, which stretches the aliasing mechanism way beyond what the internal tests do. This probably means we need more internal testing for these extreme cases. Following that, the main builders should be green.

The last item is now the YASK builder, and I'm on it. @ggorman @navjotk Now would be a good time to get another review or two in... :wink:

FabioLuporini commented on devito/ir/iet/nodes.py:
```diff
@@ -419,6 +486,29 @@ def children(self):
         return (self.nodes,)
 
+    @property
+    def functions(self):
```
I'm adding this comment here, but it applies to lots of places. I think all these methods, ``functions, free_symbols, ...`` should return tuples, not lists

FabioLuporini replied:
In fact, after skimming through the changes, it seems like there are some inconsistencies ?

navjotk replied:
This probably needs to be addressed

FabioLuporini replied:
yes

FabioLuporini commented on devito/ir/iet/visitors.py:
```diff
@@ -132,34 +131,59 @@ def _args_decl(self, args):
+                    ret.append(i.name)
+                elif i.is_TensorFunction:
+                    ret.append('%s_vec' % i.name)
+            except:
```
Something better than just except?

navjotk replied:
Should this be addressed?

FabioLuporini replied:
I don't know how

FabioLuporini replied:
I'm guessing ``AttributeError``. Let's see...

FabioLuporini commented on devito/operator.py:
```diff
@@ -96,10 +95,10 @@ def __init__(self, expressions, **kwargs):
 
         # Translate into backend-specific representation (e.g., GPU, Yask)
-        nodes = self._specialize(nodes, parameters)
+        nodes = self._specialize(nodes)
```
I'm pretty sure that the YASK backend modifies ``parameters``  in place, but I cannot find any changes to the YASK backend?

FabioLuporini commented on devito/operator.py:
```diff
@@ -109,32 +108,81 @@ def __init__(self, expressions, **kwargs):
-        self.argument_engine = ArgumentEngine(clusters.ispace, parameters,
-                                              self.dle_arguments)
+        # Insert data and pointer casts for array parameters and profiling structs
+        casts = [ArrayCast(f) for f in self.input if f.is_Tensor and f._mem_external]
```
I'm re-adding a comment that GitHub dropped after the latest changes: this should better go in ``ir/iet/scheduler`` as a ``iet_insert_casts`` free function

FabioLuporini commented on devito/operator.py:
```diff
@@ -109,32 +108,81 @@ def __init__(self, expressions, **kwargs):
 
-        parameters = self.argument_engine.arguments
+        # Filter all internally-allocated temporary `Array` types
+        # TODO: Huge hack, but to fix it, allocations need to be part of IET
```
OK, I've thought about this some more. Perhaps it would be better to let ``derive_parameters`` take an optional parameter that, if set to True, drops all local arrays? It would kinda make sense to me

FabioLuporini commented on devito/operator.py:
```diff
@@ -109,32 +108,81 @@ def __init__(self, expressions, **kwargs):
 
-        arguments, autotune = self.argument_engine.handle(**kwargs)
+        # Add in the profiler argument
+        arguments[self.profiler.name] = self.profiler.new()
```
Should we move this to ``_argument_defaults()`` ?

FabioLuporini commented on devito/ir/iet/visitors.py:
```diff
@@ -132,34 +131,59 @@ def _args_decl(self, args):
         """Convert an iterable of :class:`Argument` into cgen format."""
```
Should update the docstring?

navjotk replied:
This

FabioLuporini replied:
yes

FabioLuporini commented:
I think I've fixed the YASK backend tests. I also pleased flake8 and fixed a couple of leftover trivial bugs. Let's see what Travis has to say now...

FabioLuporini commented on devito/operator.py:
```diff
@@ -117,24 +119,66 @@ def __init__(self, expressions, **kwargs):
-            arguments = self._autotune(arguments)
+            # AT assumes and ordered dict, so let's feed it one
+            at_args = OrderedDict([(p.name, arguments[p.name]) for p in self.parameters])
+            arguments = self._autotune(at_args)
```
*looks* OK

navjotk commented on tests/test_operator.py:
```diff
@@ -588,10 +584,25 @@ def test_argument_derivation_order(self, nt=100):
         # Providing a scalar argument explicitly should override the automatically\
         # inferred
-        op_arguments = op.arguments(b=b1, time=nt - 1, time_e=nt - 2)
+        op_arguments = op.arguments(b=b1, time_e=nt - 2)
```
Why has this been changed?

FabioLuporini replied:
So Michael changed it; I **think** it's because now ``time`` aliases to ``time_e``

FabioLuporini commented:
@navjotk addressed your comments. I'll merge once formally approved.

navjotk commented:
GTG

FabioLuporini commented:
Merged.
